src/cmd/services/m3coordinator/ingest/carbon/ingest.go

```diff
@@ -148,9 +152,110 @@ type ingester struct {
 	metrics              carbonIngesterMetrics
 	tagOpts              models.TagOptions
 
+	lineResourcesPool pool.ObjectPool
+	rulesUpdatedCh    chan bool
+
+	sync.RWMutex
 	rules []ruleAndRegex
+}
 
-	lineResourcesPool pool.ObjectPool
+func (i *ingester) OnUpdate(clusterNamespaces m3.ClusterNamespaces) {
+	rules := i.regenerateIngestionRules(clusterNamespaces)
+	if rules == nil {
+		namespaces := make([]string, 0, len(clusterNamespaces))
+		for _, ns := range clusterNamespaces {
+			namespaces = append(namespaces, ns.NamespaceID().String())
+		}
+		i.logger.Warn("generated empty carbon ingestion rules from latest cluster namespaces update. leaving"+
+			" current set of rules as-is.", zap.Strings("namespaces", namespaces))
+		return
+	}
+
+	compiledRules, err := i.compileRules(*rules)
+	if err != nil {
+		i.logger.Error("failed to compile latest rules. continuing to use existing carbon ingestion "+
+			"rules", zap.Error(err))
+		return
+	}
+
+	i.Lock()
+	i.rules = compiledRules
+	i.Unlock()
+
+	if i.rulesUpdatedCh != nil {
+		i.rulesUpdatedCh <- true
+	}
+}
+
+func (i *ingester) regenerateIngestionRules(clusterNamespaces m3.ClusterNamespaces) *CarbonIngesterRules {
+	var (
+		rules = &CarbonIngesterRules{
+			Rules: i.opts.IngesterConfig.RulesOrDefault(clusterNamespaces),
+		}
+		namespacesByRetention = make(map[m3.RetentionResolution]m3.ClusterNamespace, len(clusterNamespaces))
+	)
+
+	for _, ns := range clusterNamespaces {
+		if ns.Options().Attributes().MetricsType == storagemetadata.AggregatedMetricsType {
+			namespacesByRetention[m3.RetentionResolution{
+				Resolution: ns.Options().Attributes().Resolution,
+				Retention:  ns.Options().Attributes().Retention,
+			}] = ns
+		}
+	}
+
+	// Validate rule policies.
+	for _, rule := range rules.Rules {
+		// Sort so we can detect duplicates.
+		sort.Slice(rule.Policies, func(i, j int) bool {
+			if rule.Policies[i].Resolution == rule.Policies[j].Resolution {
+				return rule.Policies[i].Retention < rule.Policies[j].Retention
+			}
+
+			return rule.Policies[i].Resolution < rule.Policies[j].Resolution
+		})
+
+		var lastPolicy config.CarbonIngesterStoragePolicyConfiguration
+		for idx, policy := range rule.Policies {
+			if idx > 0 && policy == lastPolicy {
```

Comment:
Think we're missing setting `lastPolicy` at end of each loop?

Reply:
Though I realize this isn't your code 😄  so existing bug

Reply:
Great catch. And of course, fixing that caused a test failure, lol.

Reply:
As far as I can see we only use `lastPolicy` to ensure the following policy is valid, and don't do anything with the value outside of that?

Reply:
That's correct, but we weren't even setting `lastPolicy = policy` at the end of the loop so the `lastPolicy == policy` would have never been true.